### PR TITLE
[Balance][Mystery Encounter] Balance changes and updates to various MEs

### PR DESCRIPTION
**Pull request by ben-lear**
## Balance/Development Changes

Fiery Fallout:
- Option 1 remove free QD > replace with +1 SpDef/Spd
- Option 1 and 3, change guaranteed Charcoal rewards > randomly generated Attack Type Boost item
- Option 2, the burned pokemon's ability is permanently swapped to Heatproof
- Option 3, change requirements from 2 Fire Types > 1 Fire Type OR mon with Well-Baked Body/Heatproof/Flash Fire/etc.

Vitamin Dealer:
- Pricey Deal money cost updated from 3.5x to 5x

Expert Breeder:
- Egg ratios changed from 6:1 > 4:1 (player gets more eggs)
- Egg rewards give 5 Common eggs first, then the rest of points are divided into Rare/Common eggs based on above ratio
- Guaranteed Soothe Bell in fight rewards
- Change fight levels from STRONG to AVERAGE config

Trash to Treasure:
- Black Sludge also affects shop reroll cost by 2.5x now
- Change fight level modifier from 1x to 0.5x (slightly easier fight)

Delibirdy:
- Giving a Held Item option now rewards player with Healing Charm (was Berry Pouch)
- Giving a Reviver Seed now rewards player with a Berry Pouch (was Healing Charm)

An Offer You Can't Refuse:
- Option 1 Shiny Charm replaced with new item, Silver Pokeball. Functions the same as a Golden Pokeball, but lasts 100 floors only.

Weird Dream:
- Old Gateau changed to give +20 on 3 stats: hp/speed, atk/spatk, def/spdef
- Old Gateaus are given to every mon below 570 BST on the newly transformed team
- Level decrease option lowered from 12.5% to 10%
- Add another full option to the ME
  - Player battles against what their team would have been transformed into if they had picked Option 1
  - Enemy team will have the same held items, etc. (not stealable)
  - 2 Rogue/2 Ultra/2 Great rewards (same as brutal Mysterious Challengers)

### CODE REVIEW

**Review by f-fsantos — CHANGES_REQUESTED** [review]
Frozen until my approval

**Comment by Fontbane** [comment]
On breaking changes:

1. We should use version migration system to avoid this
2. I'm not sure I like mints being tied to an ME class to begin with. Can CustomPokemonData just be moved?

**Review by CodeTappert — APPROVED** [review]

File: src/data/mystery-encounters/mystery-encounter-requirements.ts [hunk 1]
```diff
@@ -577,10 +577,10 @@ export class AbilityRequirement extends EncounterPokemonRequirement {
 
   override queryParty(partyPokemon: PlayerPokemon[]): PlayerPokemon[] {
     if (!this.invertQuery) {
-      return partyPokemon.filter((pokemon) => this.requiredAbilities.some((ability) => pokemon.getAbility().id === ability));
+      return partyPokemon.filter((pokemon) => this.requiredAbilities.some((ability) => pokemon.getAbility().id === ability || pokemon.getPassiveAbility().id === ability));
```
Comment (DayKev) [inline src/data/mystery-encounters/mystery-encounter-requirements.ts:580]: Shouldn't these be using `hasAbility(ability)`?

File: src/data/mystery-encounters/utils/encounter-pokemon-utils.ts [hunk 2]
```diff
@@ -833,3 +835,23 @@ export function isPokemonValidForEncounterOptionSelection(pokemon: Pokemon, scen
 
   return null;
 }
+
+/**
+ * Permanently overrides the ability (not passive) of a pokemon.
+ * If the pokemon is a fusion, instead overrides the fused pokemon's ability.
+ * @param pokemon
+ * @param ability
```
Comment (DayKev) [inline src/data/mystery-encounters/utils/encounter-pokemon-utils.ts:843]:
```suggestion
```

File: src/modifier/modifier.ts [hunk 3]
```diff
@@ -403,6 +403,15 @@ export abstract class LapsingPersistentModifier extends PersistentModifier {
     this.battleCount = this.maxBattles;
   }
 
+  /**
+   * Updates an existing modifier with a new `maxBattles` and `battleCount`.
+   * @param count
```
Comment (DayKev) [inline src/modifier/modifier.ts:408]:
```suggestion
```

File: src/modifier/modifier.ts [hunk 4]
```diff
@@ -3285,6 +3294,60 @@ export class ExtraModifierModifier extends PersistentModifier {
...
+  }
+
+  match(modifier: Modifier): boolean {
+    return (modifier instanceof TempExtraModifierModifier);
+  }
+
+  /**
+   * Increases the current rewards in the battle by the stackCount.
+   * @returns true if the shop reward number modifier applies successfully
+   * @param count {@linkcode Utils.IntegerHolder} that holds the resulting shop item reward count
```
Comment (DayKev) [inline src/modifier/modifier.ts:3343]:
```suggestion
   * Increases the current rewards in the battle by the `stackCount`.
   * @returns `true` if the shop reward number modifier applies successfully
   * @param count {@linkcode NumberHolder} that holds the resulting shop item reward count
```

File: src/phases/select-modifier-phase.ts [hunk 5]
```diff
@@ -260,7 +261,13 @@ export class SelectModifierPhase extends BattlePhase {
       // Otherwise, continue with custom multiplier
       multiplier = this.customModifierSettings.rerollMultiplier;
     }
-    return Math.min(Math.ceil(this.scene.currentBattle.waveIndex / 10) * baseValue * Math.pow(2, this.rerollCount) * multiplier, Number.MAX_SAFE_INTEGER);
+
+    const baseMultiplier = Math.min(Math.ceil(this.scene.currentBattle.waveIndex / 10) * baseValue * Math.pow(2, this.rerollCount) * multiplier, Number.MAX_SAFE_INTEGER);
```
Comment (DayKev) [inline src/phases/select-modifier-phase.ts:265]:
```suggestion
    const baseMultiplier = Math.min(Math.ceil(this.scene.currentBattle.waveIndex / 10) * baseValue * (this.rerollCount ** 2) * multiplier, Number.MAX_SAFE_INTEGER);
```

**Comment by DayKev** [comment]
> 2. I'm not sure I like mints being tied to an ME class to begin with. Can CustomPokemonData just be moved?

Agreed, if `CustomPokemonData` is now no longer used only by MEs then it should be moved to a more general location.

**Comment by Tempo-anon** [comment]
Closing in favor of https://github.com/pagefaultgames/pokerogue/pull/4700 which has resolved merge conflicts

**Pull request closed without merging**